nano-sync/src/circuits/mnt4/pk_tree_5.rs

```diff
@@ -0,0 +1,314 @@
+            prepare_agg_pk_commitment,
+            commit_signer_bitmap,
+            commit_agg_pk_commitment,
+            position,
+        }
+    }
+}
+
+impl ConstraintSynthesizer<MNT4Fr> for PKTree5Circuit {
+    /// This function generates the constraints for the circuit.
+    fn generate_constraints<CS: ConstraintSystem<MNT4Fr>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+        // Allocate all the constants.
+        #[allow(unused_mut)]
+        let mut cost = start_cost_analysis!(cs, || "Alloc constants");
+
+        let sum_generator_g1_var =
+            G1Gadget::alloc_constant(cs.ns(|| "alloc sum generator g1"), &sum_generator_g1_mnt6())?;
+
+        let sum_generator_g2_var =
+            G2Gadget::alloc_constant(cs.ns(|| "alloc sum generator g2"), &sum_generator_g2_mnt6())?;
+
+        let pedersen_generators_var = Vec::<G1Gadget>::alloc_constant(
+            cs.ns(|| "alloc pedersen_generators"),
+            pedersen_generators(195),
+        )?;
+
+        // Allocate all the private inputs.
+        next_cost_analysis!(cs, cost, || { "Alloc private inputs" });
+
+        let pks_var = Vec::<G2Gadget>::alloc(cs.ns(|| "alloc public keys"), || Ok(&self.pks[..]))?;
+
+        let pks_nodes_var =
+            Vec::<G1Gadget>::alloc(cs.ns(|| "alloc pks Merkle proof nodes"), || {
+                Ok(&self.pks_nodes[..])
+            })?;
+
+        let prepare_agg_pk_var =
+            G2Gadget::alloc(cs.ns(|| "alloc prepare aggregate public key"), || {
+                Ok(&self.prepare_agg_pk)
+            })?;
+
+        let commit_agg_pk_var =
+            G2Gadget::alloc(cs.ns(|| "alloc commit aggregate public key"), || {
+                Ok(&self.commit_agg_pk)
+            })?;
+
+        // Allocate all the public inputs.
+        next_cost_analysis!(cs, cost, || { "Alloc public inputs" });
+
+        let pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc public keys commitment"),
+            self.pks_commitment.as_ref(),
+        )?;
+
+        let prepare_signer_bitmap_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc prepare signer bitmap"),
+            self.prepare_signer_bitmap.as_ref(),
+        )?;
+
+        let prepare_agg_pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc prepare aggregate pk commitment"),
+            self.prepare_agg_pk_commitment.as_ref(),
+        )?;
+
+        let commit_signer_bitmap_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc commit signer bitmap"),
+            self.commit_signer_bitmap.as_ref(),
+        )?;
+
+        let commit_agg_pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc commit aggregate pk commitment"),
+            self.commit_agg_pk_commitment.as_ref(),
+        )?;
+
+        let position_var =
+            UInt8::alloc_input_vec(cs.ns(|| "alloc position"), self.position.as_ref())?
+                .pop()
+                .unwrap();
```

Comment:
We don't discard all but one elements, the vector only has one element. Since the input gadget only works with Vec<UInt8> I'm forced to use a 1-element vector here.

Reply:
Actually I was having a bug because of it and it was you who found the solution. Maybe I should write some comments explaining it.

Reply:
I would remove all these `Vec` occurrences and only pass a single `Uint8`.
The input gadget problem can be easily solved by calling:
```rust
        proof_inputs.append(&mut RecursiveInputGadget::to_field_elements::<Fr>(
            &[right_position],
        )?);
```
Note that I replaced `&right_position` by `&[right_position]`.

Reply:
Done!

Reply:
Update: This started to give problems. Whenever it verified a SNARK proof inside a circuit it would complain that the length of the verifying key is different from the length of the proof.

Reply:
That's because of `Uint8::alloc_input` vs. `Uint8::alloc_input_vec`. The first one allocates one field element per bit, the second allocates one field element for the whole Uint8.
I just made the changes myself to speed things up and will merge the code soon.